File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/observation.py
# nvdaMcpBridge domain -- observation: what a command reports having seen.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: supporting construct, the pure mappings from entity state to wire results shared by several commands.

from __future__ import annotations

from typing import TYPE_CHECKING

from .... import protocol
from .wallclock import format_wallclock

if TYPE_CHECKING:
	from .session_context import SessionContext


def speech_entries(entries: list[tuple[str, int, int, float]]) -> list[protocol.SpeechEntry]:
	return [
		protocol.SpeechEntry(
			text=text,
			index=index,
			logPosition=log_position,
			emittedAt=format_wallclock(emitted_at),
		)
		for text, index, log_position, emitted_at in entries
	]


def state_snapshot(ctx: SessionContext) -> protocol.StateResult:
	"""Raises on a browse mode outside the wire's closed set rather than publishing it."""
	state = ctx.adapter_set.state_inspector.state()
	return protocol.StateResult(
		browseMode=protocol.BrowseMode(state.browse_mode),
		speechMode=state.speech_mode,
		sleepMode=state.sleep_mode,
		inputHelp=state.input_help,
	)

File: server/domain/controllers/tools/observation.go
// screenreader-mcp domain -- the observation a mutating tool reports.
// Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
// ROLE: supporting construct; the result half press_gesture and type_text answer identically, plus the mappings that fill it.
package tools

import (
	"errors"
	"strings"

	"github.com/marlon-sousa/screen-readers-mcp/server/domain/ports"
)

// DefaultGraceMs is a heuristic from one measurement: speech ~124 ms after a keystroke, against a ~2.6 s agent round trip.
const DefaultGraceMs = 100

// DefaultTypeGraceMs is zero: with speak typed characters on, typing emits one utterance per character and none is worth waiting for.
const DefaultTypeGraceMs = 0

// An empty Speech means nothing had arrived by the stated instant, never that nothing more will.
type observation struct {
	Speech []capturedEntry `json:"speech"`
	// [SpeechFrom, SpeechTo) is half-open; SpeechTo is where to read from next.
	SpeechFrom int `json:"speechFrom"`
	SpeechTo   int `json:"speechTo"`
	// State is absent when the reader serves no state capability, which is a different answer from all fields zero.
	State *stateResult `json:"state,omitempty"`
	// Announced is omitted when none was asked for, and says the announcement was made, never that it was heard.
	Announced string `json:"announced,omitempty"`
}

func observed(o ports.Observation, announced string) observation {
	return observation{
		Speech:     capturedEntries(o.Speech),
		SpeechFrom: o.FromIndex,
		SpeechTo:   o.ToIndex,
		State:      stateSnapshot(o.State),
		Announced:  announced,
	}
}

// announcement treats absent and empty as silence, and rejects whitespace-only text.
func announcement(text string) (string, error) {
	if text == "" {
		return "", nil
	}
	if strings.TrimSpace(text) == "" {
		return "", errors.New("announce must not be whitespace only: send no announce at all to stay quiet")
	}
	return text, nil
}

// capturedEntries never returns nil, so an agent finds an empty list rather than JSON null.
func capturedEntries(entries []ports.SpeechEntry) []capturedEntry {
	mapped := make([]capturedEntry, 0, len(entries))
	for _, entry := range entries {
		mapped = append(mapped, capturedEntry{
			Text:        entry.Text,
			Index:       entry.Index,
			LogPosition: entry.LogPosition,
			EmittedAt:   entry.EmittedAt,
		})
	}
	return mapped
}

func stateSnapshot(state *ports.ReaderState) *stateResult {
	if state == nil {
		return nil
	}
	return &stateResult{
		BrowseMode: state.BrowseMode,
		SpeechMode: state.SpeechMode,
		SleepMode:  state.SleepMode,
		InputHelp:  state.InputHelp,
	}
}
